### PR TITLE
Feature/LJI-19-claims-filters

### PR DESCRIPTION
**Pull request by anddevdev**
This PR introduces the following features for the Claims:

1. Filtering claims:
- Allows filtering of claims by claim number, consumer (employee), employer (company), and status.
- Multiple filters can be applied simultaneously (filter by company and status at the same time).

2. Sorting claims:
- Enables sorting of claims by claim number, consumer, employer, date of service, plan, amount, and status(Ascending and descending orders)
- Default sorting is enabled by date of service in descending order (newest claims first).

3. Pagination 
- Default pagination is set to 10 entities per page

4. New endpoint:
- Added a new endpoint /v1/claims/filtered for handling filtered, sorted, and paginated requests.
- Added new requests to Claims collection in Postman to test the added features.

### CODE REVIEW

File: src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java [hunk 1]
```diff
@@ -22,6 +38,31 @@ public ResponseEntity<List<ClaimDTO>> getAllClaims() {
...
+        .orElse(new FilterCriteria());
+    SortCriteria sortCriteria = Optional.ofNullable(criteria.getSortCriteria())
+        .orElseGet(() -> {
+          SortCriteria defaultSort = new SortCriteria();
+          defaultSort.setField("dateOfService");
+          defaultSort.setDirection(Sort.Direction.DESC);
+          return defaultSort;
+        });
+
+    int pageSize = pageable.getPageSize() == 20 ? 10 : pageable.getPageSize();
```
Comment (anton-kolyaev) [inline src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java:56]: (⊙ _ ⊙ )

Comment (anddevdev) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java]: Deleted, just set the page size to 10

File: src/main/java/com/coherentsolutions/pot/insurance/service/ClaimService.java [hunk 2]
```diff
@@ -20,6 +27,33 @@ public class ClaimService {
...
+    Sort sort = Sort.by(sortCriteria.getDirection(), sortCriteria.getField());
+    PageRequest pageRequest = PageRequest.of(page, size, sort);
+
+    return claimRepository.findAll(spec, pageRequest).map(ClaimMapper.INSTANCE::entityToDto);
+  }
+
+  private Specification<ClaimEntity> buildSpecification(FilterCriteria filterCriteria) {
+    Specification<ClaimEntity> spec = Specification.where(null);
+
+    if (filterCriteria.getClaimNumber() != null && !filterCriteria.getClaimNumber().isEmpty()) {
```
Comment (anton-kolyaev) [inline src/main/java/com/coherentsolutions/pot/insurance/service/ClaimService.java:41]: can we do any refactoring to make it "prettier"?

**Review by anton-kolyaev — CHANGES_REQUESTED** [review]

File: src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java [hunk 3]
```diff
@@ -22,6 +38,31 @@ public ResponseEntity<List<ClaimDTO>> getAllClaims() {
...
+    Sort sort = Sort.by(sortCriteria.getDirection(), sortCriteria.getField());
+    PageRequest pageRequest = PageRequest.of(pageable.getPageNumber(), pageSize, sort);
+
+    Page<ClaimDTO> claimsPage = claimService.getFilteredSortedClaims(filterCriteria, sortCriteria, pageRequest.getPageNumber(), pageRequest.getPageSize());
+
+    return ResponseEntity.ok(claimsPage);
```
[the remark this answers is not in the document]
Comment (anddevdev) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java]: Done

File: src/main/java/com/coherentsolutions/pot/insurance/specifications/SortCriteria.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.coherentsolutions.pot.insurance.specifications;
+
+import lombok.Data;
+import org.springdoc.api.annotations.ParameterObject;
+import org.springframework.data.domain.Sort;
+
+@Data
+@ParameterObject
+public class SortCriteria {
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/specifications/SortCriteria.java:9]: Could we keep sorting in the Pageable object and do not create our custom class?

File: src/main/java/com/coherentsolutions/pot/insurance/specifications/FilterAndSortCriteria.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.coherentsolutions.pot.insurance.specifications;
+
+import lombok.Data;
+import org.springdoc.api.annotations.ParameterObject;
+
+
+@Data
+@ParameterObject
+public class FilterAndSortCriteria {
+  private FilterCriteria filterCriteria;
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/specifications/FilterAndSortCriteria.java:10]: What if make it more abstract to re-use it with other controllers?
I mean: receive Filters as a list with the field name and value of this field?

File: src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java [hunk 6]
```diff
@@ -18,38 +32,71 @@ public class ClaimController {
...
+    return claimService.getAllClaims();
+  }
+
+  @GetMapping("/filtered")
+  @ResponseStatus(HttpStatus.OK)
+  public Page<ClaimDTO> getFilteredSortedClaims(
+      @ParameterObject FilterAndSortCriteria criteria,
+      @ParameterObject Pageable pageable) {
+
+    FilterCriteria filterCriteria;
```
Comment (SavaLX) [inline src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java:46]: Maybe it's better to move this logic to the service layer, what do you think?

File: src/main/java/com/coherentsolutions/pot/insurance/specifications/FilterAndSortCriteria.java (hunk 5 above)
Comment (SavaLX) [reply src/main/java/com/coherentsolutions/pot/insurance/specifications/FilterAndSortCriteria.java:10]: Let's leave it for now

File: src/main/java/com/coherentsolutions/pot/insurance/specifications/SortCriteria.java (hunk 4 above)
Comment (anddevdev) [reply src/main/java/com/coherentsolutions/pot/insurance/specifications/SortCriteria.java]: Done

File: src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java (hunk 6 above)
Comment (anddevdev) [reply src/main/java/com/coherentsolutions/pot/insurance/controller/ClaimController.java]: Done

File: src/main/java/com/coherentsolutions/pot/insurance/service/ClaimService.java [hunk 7]
```diff
@@ -20,6 +27,42 @@ public class ClaimService {
...
+
+    if (!pageable.getSort().isSorted()) {
+      pageable = PageRequest.of(pageable.getPageNumber(), pageable.getPageSize(), defaultSort);
+    }
+
+    Specification<ClaimEntity> spec = buildSpecification(claimFilterCriteria);
+    return claimRepository.findAll(spec, pageable).map(ClaimMapper.INSTANCE::entityToDto);
+  }
+
+  private Specification<ClaimEntity> buildSpecification(ClaimFilterCriteria claimFilterCriteria) {
```
[the remark this answers is not in the document]
Comment (anddevdev) [reply src/main/java/com/coherentsolutions/pot/insurance/service/ClaimService.java]: Done

**Review by SavaLX — APPROVED** [review]

**Review by anton-kolyaev — DISMISSED** [review]
Add integration test(s) please

**Review by NormantasN — APPROVED** [review]

**Pull request merged**